### PR TITLE
test: warning behaviour for unset/set `TARGETOS`,`TARGETARCH`,`TARGETPLATFORM`

### PR DESCRIPTION
**Pull request by flouthoc**
Test buildah's warning behaviors for unset or set `TARGETOS`, `TARGETARCH`, `TARGETPLATFORM` or `TARGETVARIANT` args when set or unset directly or indirectly by the user.

Adds missing test for: https://github.com/containers/buildah/pull/4295

### CODE REVIEW

**Comment by flouthoc** [comment]
@lucacome @rhatdan @nalind @containers/buildah-maintainers PTAL

**Comment by rhatdan** [comment]
LGTM
@edsantiago PTAL

**Review by edsantiago — CHANGES_REQUESTED** [review]
Don't merge yet please, I have a few concerns. Am looking into ways to confirm and address.

File: tests/bud.bats
Comment (edsantiago) [inline tests/bud.bats:41]: A lot of duplication

File: tests/bud.bats
Comment (edsantiago) [inline tests/bud.bats:39]: Too explicit. It's much better to have a looser test here, such that any warning will cause a test failure.

File: tests/bud.bats
Comment (edsantiago) [inline tests/bud.bats:78]: lots of duplication

**Review by edsantiago — CHANGES_REQUESTED** [review]
Here's a shorter, cleaner version that uses only one containerfile for all tests, and is more careful about checking for warnings:
```bash
@test "build test warning for preconfigured TARGETARCH, TARGETOS, TARGETPLATFORM or TARGETVARIANT" {
  _prefetch alpine
  containerfile=${TEST_SCRATCH_DIR}/Containerfile

  # buildah should not warn when any of these is indirectly set by cli options
  local -a checkvars=(ARCH OS PLATFORM VARIANT)
  echo "FROM alpine" >$containerfile
  for var in "${checkvars[@]}"; do
    echo "ARG TARGET${var}" >>$containerfile
  done
  echo "COPY . ." >>$containerfile

  # With explicit and full --platform, there should be no warnings
  run_buildah build $WITH_POLICY_JSON --platform linux/amd64/v2 -t source -f $containerfile
  assert "$output" !~ "missing .* build argument" \
         "With explicit --platform, buildah should not warn"

  # Without --platform, we want one warning for each variable
  run_buildah build $WITH_POLICY_JSON -t source -f $containerfile
#  for var in "${checkvars[@]}"; do
  for i in $(seq 0 3); do
    var="TARGET${checkvars[$i]}"
    # The '..'s cover backslash-quote: "missing \"TARGETFOO\" ..."
    assert "${lines[$i]}" =~ "missing ..${var}.. build argument" \
           "buildah should warn about undefined ${var}"
  done
}
```
Incidentally, is `--os linux` or `--os linux/amd64` supposed to shutup the `missing TARGETOS` warning? Because it doesn't. I gave up trying to add tests for that.

**Comment by flouthoc** [comment]
Thanks @edsantiago for the patch.

> Incidentally, is --os linux or --os linux/amd64 supposed to shutup the missing TARGETOS warning? Because it doesn't. I gave up trying to add tests for that.

`--os` doesn't do but `--platform` does, I think `--os` needs to be fixed, I'll open a new PR fixing `--os` and will also amend the test then.

**Comment by edsantiago** [comment]
Thank you!

/lgtm

**Review by edsantiago — APPROVED** [review]

**Pull request merged**
